"""Frozen TLE device functions extracted without changing bodies.
Unchanged original source and licensing are retained under tle/.
"""
import triton
import triton.language as tl
from triton.experimental import tle
from triton.language.extra import libdevice
from rms_norm_w8a16_fp8 import tle_ext as ext
@triton.jit
def prev_multiple_of(a, b):
    return tl.cdiv(a, b) * b - b

@triton.jit(do_not_specialize=['eps'])
def rms_norm_fp8_w8a16_kernel(out_ptr, in_ptr, w_ptr, w_scale_ptr, N, eps, BLOCK_SIZE: tl.constexpr, GROUP_SIZE: tl.constexpr):
    pid = tl.program_id(0)
    cols = tl.arange(0, BLOCK_SIZE)
    mask = cols < N
    x = tl.load(in_ptr + pid * N + cols, mask=mask, other=0.0).to(tl.float32)
    var = tl.sum(x * x, axis=0) / N
    rrms = 1 / tl.sqrt(var + eps)
    group_ids = cols // GROUP_SIZE
    w = tl.load(w_ptr + cols, mask=mask, other=0.0).to(tl.float32)
    w_scale = tl.load(w_scale_ptr + group_ids, mask=mask, other=0.0).to(tl.float32)
    y = x * rrms * w * w_scale
    tl.store(out_ptr + pid * N + cols, y, mask=mask)

@triton.jit(do_not_specialize=['eps'])
def rms_norm_fp8_w8a16_grouped_kernel(out_ptr, in_ptr, w_ptr, w_scale_ptr, N, eps, GROUP_SIZE: tl.constexpr, NUM_GROUPS: tl.constexpr):
    pid = tl.program_id(0)
    groups = tl.arange(0, NUM_GROUPS)
    cols = tl.arange(0, GROUP_SIZE)
    offsets = groups[:, None] * GROUP_SIZE + cols[None, :]
    mask = offsets < N
    x = tl.load(in_ptr + pid * N + offsets, mask=mask, other=0.0).to(tl.float32)
    var = tl.sum(x * x) / N
    rrms = 1 / tl.sqrt(var + eps)
    w = tl.load(w_ptr + offsets, mask=mask, other=0.0).to(tl.float32)
    w_scale = tl.load(w_scale_ptr + groups).to(tl.float32)[:, None]
    y = x * rrms * w * w_scale
    tl.store(out_ptr + pid * N + offsets, y, mask=mask)

@triton.jit(do_not_specialize=['eps'])
def rms_norm_fp8_w8a16_loop_kernel(out_ptr, in_ptr, w_ptr, w_scale_ptr, N, eps, TILE_N: tl.constexpr, GROUP_SIZE: tl.constexpr):
    pid = ext.program_id(0)
    acc = tl.zeros((TILE_N,), dtype=tl.float32)
    num_steps = tl.cdiv(N, TILE_N)
    for step in range(0, num_steps - 1):
        start_n = step * TILE_N
        n_offsets = start_n + tl.arange(0, TILE_N)
        x = tl.load(in_ptr + pid * N + n_offsets).to(tl.float32)
        acc += x * x
    start_n = (num_steps - 1) * TILE_N
    n_offsets = start_n + tl.arange(0, TILE_N)
    mask = n_offsets < N
    x = tl.load(in_ptr + pid * N + n_offsets, mask=mask, other=0.0).to(tl.float32)
    acc += x * x
    var = tl.sum(acc) / N
    rrms = 1 / tl.sqrt(var + eps)
    prev_multiple = prev_multiple_of(N, TILE_N)
    for start_n in range(0, TILE_N, TILE_N):
        n_offsets = prev_multiple - start_n + tl.arange(0, TILE_N)
        mask = n_offsets < N
        x = tl.load(in_ptr + pid * N + n_offsets, mask=mask, other=0.0, eviction_policy='evict_first').to(tl.float32)
        group_ids = n_offsets // GROUP_SIZE
        w = tl.load(w_ptr + n_offsets, mask=mask, other=0.0).to(tl.float32)
        w_scale = tl.load(w_scale_ptr + group_ids, mask=mask, other=0.0).to(tl.float32)
        y = x * rrms * w * w_scale
        tl.store(out_ptr + pid * N + n_offsets, y, mask=mask)
    for start_n in range(TILE_N, N, TILE_N):
        n_offsets = prev_multiple - start_n + tl.arange(0, TILE_N)
        x = tl.load(in_ptr + pid * N + n_offsets, eviction_policy='evict_first').to(tl.float32)
        group_ids = n_offsets // GROUP_SIZE
        w = tl.load(w_ptr + n_offsets).to(tl.float32)
        w_scale = tl.load(w_scale_ptr + group_ids).to(tl.float32)
        y = x * rrms * w * w_scale
        tl.store(out_ptr + pid * N + n_offsets, y)
